Blueprints: Add import Wizard (HMS-3690)

avitova opened the pull request:
Sooo...I think we are ready for a discussion on what we want to export AND import, since this is ready. :)

avitova commented on src/Components/Blueprints/ImportBlueprintModal.tsx:
```diff
@@ -96,15 +102,25 @@ export const ImportBlueprintModal: React.FunctionComponent<
               <HelperTextItem variant={isRejected ? 'error' : 'default'}>
                 {isRejected
-                  ? 'Must be a JSON file no larger than 1 KB'
+                  ? 'Must be a JSON file no larger than 6 KB'
```
In case we want to export even the gpg key and overall everything, 1 KB does not seem to be enough.

lucasgarfield commented:
Should this be working, @avitova? When I tried importing a blueprint, I just got kicked back to the table.

avitova commented:
@lucasgarfield Yeah, you are right, let me correct this:)

amirfefer reviewed (commented):
Thanks @avitova, this looks and works great! 👍 
I have a suggestion regarding the json parsing

avitova commented on src/Components/CreateImageWizardV2/ImportImageWizard.tsx:
```diff
@@ -0,0 +1,48 @@
+  useEffect(() => {
+    if (blueprint) {
+      try {
+        const importedBlueprint: BlueprintResponse = JSON.parse(blueprint);
+        const importBlueprintState = mapRequestToState(importedBlueprint);
```
I agree, thank you for this idea. :)

avitova commented:
I addressed your comments @amirfefer :) Ready for re-review.

amirfefer commented on src/Components/Blueprints/ImportBlueprintModal.tsx:
```diff
@@ -87,23 +109,35 @@ export const ImportBlueprintModal: React.FunctionComponent<
         </FormGroup>
         <ActionGroup>
-          <Button type="button">Review and finish</Button>
+          <Button
```
can you please add `ouid` prop for testing?

amirfefer reviewed (commented):
Thanks @avitova, tested a full export-import scenario, and it works well!  👍 
I left a few minor inline comments, mainly for wrong casting.

This feature is guarded by a feature flag, so I'm ok to add unit  tests as follow-up. react testing library has `upload` API for mimicking upload input, like [here](https://github.com/osbuild/image-builder-frontend/pull/1960/files#diff-f3153ea47023210c886d905eabef84599e8b22b39c5abb7fdaad04af76dca81cR98-R102)

amirfefer commented on src/Components/CreateImageWizardV2/ImportImageWizard.tsx:
```diff
@@ -0,0 +1,34 @@
+      );
+    }
+  }, [blueprint, dispatch]);
+  return <CreateImageWizard startStepIndex={1} />;
```
This is the reason why this fails to build, `startStepIndex` prop is no longer exists, and doesn't needed here, the default is 1 anyway.

regexowl commented on src/Components/Blueprints/ImportBlueprintModal.tsx:
```diff
@@ -49,17 +60,28 @@ export const ImportBlueprintModal: React.FunctionComponent<
+      setImportedBlueprint(importBlueprintState);
+      setJsonContent(value);
+    } catch (error) {
+      console.log(error);
```
I don't think we should be logging errors to the console as those have very little value to user.

Since this would be an error with the blueprint format, would a toast notification like "Failed to parse the blueprint" or something similar work?

avitova replied:
Oh yeah, you are right. This was a leftover from testing. :eyes:

regexowl commented on src/Components/Blueprints/ImportBlueprintModal.tsx:
```diff
@@ -87,23 +109,36 @@ export const ImportBlueprintModal: React.FunctionComponent<
+            }
+            ouiaId="import-blueprint-finish"
+          >
+            Review and finish
```
More of a question - do we want to drop the user on the first step or on the review step after importing a blueprint?

Dropping them directly on the Review with all the information prefilled makes more sense to me as it saves the user the work of clicking through the Wizard. Especially since the left side navigation is disabled so there's no quick way to get to the review.

(also I thought the importing doesn't work for a moment, because I didn't notice that there are already some preselected values in the Wizard :sweat_smile:)

What do you think? @avitova @lucasgarfield

avitova replied:
We discussed this with Kelsea on one of the UI meetings. It makes sense to drop them on the first page. In the future, we will possibly omit some of the data (and also some of the mandatory data) that we allow to export/import, so users would have to go through the wizard anyways:)

regexowl replied:
Got it, makes sense!

regexowl commented on src/Components/Blueprints/ImportBlueprintModal.tsx:
```diff
@@ -87,23 +109,36 @@ export const ImportBlueprintModal: React.FunctionComponent<
+            ouiaId="import-blueprint-finish"
+          >
+            Review and finish
+          </Button>
```
When I tried to import a compose request, it didn't work which makes sense. The request is missing values that blueprint needs to have.

But the failure currently crashes the service and leaves the user stuck on the `import` path which means that even refreshing the page will not help.

![image](https://github.com/osbuild/image-builder-frontend/assets/49452678/e5853b6b-da7a-40f6-ac48-43baceacfff1)

regexowl reviewed (commented):
Thank you for working on this! Added some comments

avitova commented:
So I added a test and rebased. The tests mostly cover which files should be accepted and which should not. It is quite plain, but I think that we can go with this for a while.

regexowl commented:
@avitova Looks great! Can you please resolve the conflict and rebase?

avitova commented:
@regexowl rebased:)

regexowl commented:
/retest

regexowl reviewed (approved):
Excellent, thank you! Looks and works great :rocket:

regexowl commented:
/retest

regexowl commented:
/retest

regexowl commented:
/retest

regexowl commented:
/retest

avitova commented:
:partying_face: Thank you!
